File: app/src/main/java/uk/co/richyhbm/coinbag/balances/BlockChainInfo.java
package uk.co.richyhbm.coinbag.balances;


import android.util.Log;

import java.io.IOException;
import java.math.RoundingMode;
import java.text.DecimalFormat;

import okhttp3.Request;
import okhttp3.Response;

//Balance fetcher for BlockChain Info
public class BlockChainInfo extends Balance {
    public BlockChainInfo() {
        super(10);
    }

    @Override
    protected double getBalance(String address) throws IOException {
        //Fetches the address balance and returns it in BTC
        Request req = new Request.Builder().url("https://blockchain.info/q/addressbalance/" + address).build();
        Response res = client.newCall(req).execute();
        String s = res.body().string();
        try {
            Double satoshis = Double.parseDouble(s);
            Double btc = satoshis / 100000000.0;

            DecimalFormat df = new DecimalFormat("#.########");
            df.setRoundingMode(RoundingMode.CEILING);
            return Double.parseDouble(df.format(btc));
        }catch(ArithmeticException ae) {
            ae.printStackTrace();
            return -1;
        }
    }
}

File: app/src/main/java/uk/co/richyhbm/coinbag/requests/BlockchainInfo.kt
package uk.co.richyhbm.coinbag.requests

import retrofit2.Call
import retrofit2.Retrofit
import retrofit2.converter.scalars.ScalarsConverterFactory
import retrofit2.http.GET
import retrofit2.http.Path

object BlockchainInfo {
    interface BlockchainInfoService {
        @GET("/q/addressbalance/{address}")
        fun addressBalance(@Path("address") address: String): Call<String?>
    }

    fun getBalance(address:String) : Double {
        val retrofit = Retrofit.Builder()
                .baseUrl("https://blockchain.info")
                .addConverterFactory(ScalarsConverterFactory.create())
                .build()

        val call = retrofit.create<BlockchainInfo.BlockchainInfoService>(BlockchainInfo.BlockchainInfoService::class.java).addressBalance(address)
        val balance = call.execute().body()

        if(balance == null) {
            return 0.0
        } else {
            //Returned amount is satoshi, turn into bitcoin
            return balance.toDouble() / 100000000
        }
    }
}